crates/ruff/src/jupyter/notebook.rs

```diff
@@ -0,0 +1,291 @@
+                        if (ruff_rustpython::tokenize(&contents))
+                            .last()
+                            .map_or(true, std::result::Result::is_err)
+                        {
+                            Diagnostic::new(
+                                SyntaxError {
+                                    message: format!(
+                                        "A Jupyter Notebook (.{JUPYTER_NOTEBOOK_EXT}) must internally be JSON, \
+                                but this file isn't valid JSON: {err}"
+                                    ),
+                                },
+                                Range::default(),
+                            )
+                        } else {
+                            Diagnostic::new(
+                                SyntaxError {
+                                    message: format!(
+                                        "Expected a Jupyter Notebook (.{JUPYTER_NOTEBOOK_EXT} extension), \
+                                    which must be internally stored as JSON, \
+                                    but found a Python source file: {err}"
+                                    ),
+                                },
+                                Range::default(),
+                            )
+                        }
+                    }
+                    Category::Data => {
+                        // We could try to read the schema version here but if this fails it's
+                        // a bug anyway
+                        Diagnostic::new(
+                            SyntaxError {
+                                message: format!(
+                                    "This file does not match the schema expected of Jupyter Notebooks: {err}"
+                                ),
+                            },
+                            Range::default(),
+                        )
+                    }
+                }
+            }));
+        }
+    };
+
+    // v4 is what everybody uses
+    if notebook.nbformat != 4 {
+        // bail because we should have already failed at the json schema stage
+        return Err(Box::new(Diagnostic::new(
+            SyntaxError {
+                message: format!(
+                    "Expected Jupyter Notebook format 4, found {}",
+                    notebook.nbformat
+                ),
+            },
+            Range::default(),
+        )));
+    }
+    if !notebook
+        .metadata
+        .language_info
+        .as_ref()
+        .map_or(true, |language| language.name == "python")
+    {
+        debug!(
+            "Skipping {} because it's not a python notebook",
+            path.display()
+        );
+        return Ok(None);
+    }
+
+    Ok(Some(notebook))
+}
+
+/// Concatenates all cells into a single virtual file and builds an index that maps the content
+/// to notebook cell locations
+pub fn concat_notebook(notebook: &JupyterNotebook) -> (String, JupyterIndex) {
```

Comment:
Nit: I expected this function to concat two notebooks. What do you think of adding a `index()` function on `JupyterNotebook`?

Reply:
good point, i renamed it `index_notebook`